File: 45.跳跃游戏-ii.cpp
/*
 * @lc app=leetcode.cn id=45 lang=cpp
 *
 * [45] 跳跃游戏 II
 */
#include<vector>
using namespace std;
// @lc code=start
class Solution {
public:
    int jump(vector<int> &nums) {
        int len = nums.size();
        int count = 0;
        int begin = 0;
        while (begin <= len - 1) {
            if (begin == len - 1) {
                break;
            } else if (begin + nums[begin] >= len - 1) {
                count++;
                break;
            } else {
                int temp = 0;
                int newBegin = begin;
                for (int i = 1; i <= nums[begin]; i++) {
                    if (temp <= nums[begin + i] + i) {
                        temp = nums[begin + i] + i;
                        newBegin = begin + i;
                    }
                }
                count++;
                begin = newBegin;
            }
        }
        return count;
    }
};
// @lc code=end

File: 45.跳跃游戏-ii.rs
/*
 * @lc app=leetcode.cn id=45 lang=rust
 *
 * [45] 跳跃游戏 II
 */

// @lc code=start
impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        let mut count = 0;
        let mut begin = 0;
        while begin <= len - 1 {
            if begin == len - 1 {
                break;
            } else if begin + nums[begin] as usize >= len - 1 {
                count += 1;
                break;
            } else {
                let mut temp = 0;
                let mut new_begin = begin;
                for i in 1..=nums[begin] as usize {
                    if temp <= nums[begin + i] + i as i32 {
                        temp = nums[begin + i] + i as i32;
                        new_begin = begin + i;
                    }
                }
                count += 1;
                begin = new_begin;
            }
        }
        count
    }
}
// @lc code=end

